### DISCUSSION TITLE
The best way to insert multidimensional numeric array

### DISCUSSION THREAD

**Discussion by akatasonov** [jackc/pgx#1891] [Q&A] [answered]
I have a field in the database that is NUMERIC[][] and in the Go code it is represented as [][]string. But when I try to insert the values using the conn.CopyFrom and pgx.CopyFromSlice methods I get the error below.
```
COPY from stdin failed: unable to encode [][]string{[]string{"0.00342100", "0.00000000"}, []string{"0.00342000", "100.42000000"}} into binary format for _numeric (OID 1231): cannot find encode plan (SQLSTATE 57014)
```
What is the best way to insert a multi-dimensional NUMERIC array?

**Comment by jackc** [comment, marked as answer]
In PostgreSQL there are two formats that most values can be encoded, text and binary. For normal queries, such as produced by `Query` and `Exec`, PostgreSQL allows choosing the format per value. When pgx detects your underlying data is a string it simply passes it to PostgreSQL as-is.

However, the COPY protocol requires all values to be encoded in the same format. pgx's `CopyFrom` only supports the binary format. It already tried to parse strings to values that can then be encoded into the binary format. But it didn't handle types such as `[]string` that aren't strings but that can be encoded as text and then parsed and converted to binary. It now does.

Try master and it should work.

**Reply by akatasonov** [reply]
It is working now, thanks a lot for your work!
